### PR TITLE
Frozen state of `working` at review-time on Tuesday

### PR DESCRIPTION
**Pull request by tbrent**

### CODE REVIEW

File: contracts/Proposal.sol
Comment (jeremyschlatter) [inline contracts/Proposal.sol:26]: I have some inclination for this to be a struct rather than a full contract. I need to think through it more, though.

Did you start with this being a struct instead?

File: contracts/Proposal.sol
Comment (jeremyschlatter) [inline contracts/Proposal.sol:37]: I think this name makes more sense in the singular form: Statuses -> Status

File: contracts/Proposal.sol
Comment (jeremyschlatter) [inline contracts/Proposal.sol:56]: I would prefer writing these vertically. Easier to scan and see that they are matched up correctly.

File: contracts/Proposal.sol
Comment (jeremyschlatter) [inline contracts/Proposal.sol:93]: Probably want .add and .sub from SafeMath here to catch overflows and underflows.

File: contracts/Proposal.sol
Comment (jeremyschlatter) [inline contracts/Proposal.sol:94]: It seems to me like this might prevent changes that remove a token type from the basket, unless you leave one of them in the backing at the end.

That may not be true -- there might be another way to do that -- but it also seems like a bit of a weird check, given that it will only catch reducing the supply to exactly zero (it can't catch underflows b/c this is an unsigned int).

File: contracts/ownership/Ownable.sol
Comment (jeremyschlatter) [inline contracts/ownership/Ownable.sol:59]: copy-paste error: "new owner is 0 address" is not the right message here.

Could be "unauthorized" or similar instead.

File: contracts/Manager.sol
Comment (jeremyschlatter) [inline contracts/Manager.sol:157]: I would much prefer to have `_notPaused`, `_onlyWhitelist`, and `_onlyOperator` written inline in their respective modifiers above rather than be pulled out to separate functions like this, unless there is an important reason to do it this way.

File: contracts/Manager.sol
Comment (jeremyschlatter) [inline contracts/Manager.sol:407]: Could be nice to expose this publicly. In that world I might extract a public function called `isFullyCollateralized`, that returns a bool, and then have this function just be `assert(isFullyCollateralized())`.

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:157]: Writing it this way reduces bytecode, because the way modifiers actually work is they are in-lined in every place they are used.

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:407]: Great idea.

File: contracts/Proposal.sol
Comment (tbrent) [reply contracts/Proposal.sol:93]: Good catch.

File: contracts/Manager.sol
Comment (jeremyschlatter) [reply contracts/Manager.sol:157]: I see. Alas. Can you add a comment to that effect?

File: contracts/Proposal.sol
Comment (tbrent) [reply contracts/Proposal.sol:26]: Yep, started as a struct. I forget the precise reason, but the struct approach didn't play nice with having array subfields.

Comment (tbrent) [reply contracts/Proposal.sol:26]: Actually, maybe it was Basket as a subfield that was a problem. I forget.

File: contracts/Manager.sol
Comment (jeremyschlatter) [inline contracts/Manager.sol:196]: Also `require(_tokens.length == _amountsIn.length, "token quantities mismatched");` ?

Comment (tbrent) [reply contracts/Manager.sol:196]: Nice catch. Tokens was deleted and re-added and this was dropped.

File: contracts/Manager.sol
Comment (jeremyschlatter) [inline contracts/Manager.sol:277]: `if (quantitiesIn[i] > 0) {` ?

I realize that this would add to the bytecode length. Might be fine to remove, but seems a bit unfortunate if so. Should at least make sure that the backing tokens all allow transfers of size zero.

**Comment by jeremyschlatter** [comment]
I just finished reading the contracts. Their high-level design seems like it basically makes sense. I'll think about it for a bit and see if I have any thoughts about improving them.

**Comment by jeremyschlatter** [comment]
After having thought about it more:

- Something feels a bit too complicated with `Basket` and `Proposal`. I won't spend more time thinking about them until after I see @fiddlemath's refactor of `Proposal`, though. `Manager` and `Vault` both seem like basically necessary components.

- This design seems like it gives quite a bit of flexibility to the owner. There may be places where it makes sense to reduce the owner's power in exchange for increasing the trustworthiness of the system. In particular, the current design seems like it allows the owner to empty the vault within 24 hours. I would put a high priority on trying to increase that particular delay, if possible. For example, maybe the owner could commit to a new portfolio balance further out -- say a week or more -- but still allow incremental portfolio shifts toward that commitment with a shorter delay per shift.

**Comment by tbrent** [comment]
Nevin has thought through the 24 hour delay and we're okay with it currently. The rationale is that the market is able to see any malicious intentions to drain the vault and redeem their RSV. And there is a tradeoff against how quickly the system can react to things like a collateral asset failing. 

But, I do think it makes sense to make the delay flexible as an auth parameter. It doesn't make sense for it to be fixed.

**Comment by tbrent** [comment]
One note: It's already the case that owner can empty the vault regardless of the delay setting. Since Owner can change the Vault's `manager` variable, they can give themselves the ability to call `batchWithdrawTokens`. The idea with the auth permissions we've landed on is not to completely prevent a single account from being able to attack the system, but to make it so that the single account that can is used infrequently and separate from the account that is used in day-to-day operations.

File: LICENSE
Comment (puremachinery) [inline LICENSE:1]: Sanity check: Who got approval from Gertler that this license is okay? Where do we have it written down?

File: Makefile
Comment (puremachinery) [inline Makefile:6]: Nit pick: Most of the Makefile has a newline after the colon. Consistency is good.

File: README.md
Comment (puremachinery) [inline README.md:2]: Low priority but making this more informative after we're done with the critical stuff would be good.

File: contracts/Basket.sol
Comment (puremachinery) [inline contracts/Basket.sol:18]: Rename to backingByAddress or weightsByAddress? Good when names indicate both aspects of a map.

File: contracts/Basket.sol
Comment (puremachinery) [inline contracts/Basket.sol:33]: When would we want this to be something other than 18? Probably constraints like ">=0 && <=18" could go here to prevent simple mistakes, or whatever other constraints make sense given the question of when it might be not-18.

File: contracts/Basket.sol
Comment (puremachinery) [inline contracts/Basket.sol:6]: Not sure "front token" is the clearest term here. If we keep it, let's add a definition to the design docs and to a code comment to clarify, since it appears not to be a commonly used term nor mentioned anywhere in our docs thus far. Also specifying how this is different from "collateral token".

File: contracts/Basket.sol
Comment (puremachinery) [inline contracts/Basket.sol:15]: Why index by integer rather than map from token three letter code or something else that provides meaning? Maybe I'm misunderstanding the tokens and backing arrays.

File: contracts/Basket.sol
Comment (puremachinery) [inline contracts/Basket.sol:52]: More clarification of quantitiesRequired vs newQuantitiesRequired please. At minimum in a conversation with Zera, but probably in the comments and design docs.

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:38]: Is this going to stay true? That anyone can propose a change to the backing. I recall Taylor mentioning we should possibly restrict proposal changes to just us but don't remember what the resolution of that conversation was.

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:74]: Rename `proposals` to something conveying its map nature. What is the uint256 here? Proposal id?

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:79]: Rename to whitelistsByAddress or other consistent map-naming convention.

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:118]: I saw we're using leading underscores to denote function parameters or internal functions elsewhere in this code, but am not sure why there's a trailing underscore here. Move it to the front of the word? Or let me know what it means if different.

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:167]: Remove unused or commented out code.

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:178]: Remove.

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:220]: Nit: Better to remove all instances of "Importantly" and "Most importantly" from the comments. These are two sentence comments; we don't need to call out that one of the two sentences is Very Important.

"The road to hell is paved with adverbs." - Stephen King
https://www.writingforward.com/writing-tips/avoid-adverbs

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:222]: much -> many

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:234]: Do quantitiesIn and quantitiesOut not need to be the same length with one another, and also not constrained by tokens.length?

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:256]: ///?

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:299]: Is the basket not required to pause or perform other functions? If not, why not? If yes, add this require line to those other functions. And when would we set the basket address to the zero address?

File: contracts/Manager.sol
Comment (puremachinery) [inline contracts/Manager.sol:385]: Remove commented out code.

File: contracts/Proposal.sol
Comment (puremachinery) [inline contracts/Proposal.sol:19]: These are different enough types of proposals that maybe we should separate them out. `BasketProposal` and `ExchangeProposal` with a separate `Proposal` they can both extend for the overlap between them?

File: contracts/ownership/Ownable.sol
Comment (puremachinery) [inline contracts/ownership/Ownable.sol:67]: When would we want to do this?

File: contracts/rsv/Reserve.sol
Comment (puremachinery) [inline contracts/rsv/Reserve.sol:72]: I would like to determine whether this will fail here or later before we approve of this code. :)

File: contracts/rsv/Reserve.sol
Comment (puremachinery) [inline contracts/rsv/Reserve.sol:125]: Perhaps we can add an upgrade bool to set and a require here as an extra precaution? Maybe overkill / not adding value though. Not sure.

File: contracts/rsv/Reserve.sol
Comment (puremachinery) [inline contracts/rsv/Reserve.sol:194]: Why check this when also checking that it's been frozen for at least 4 weeks? Seems like a subset of the later check.

**Review by puremachinery — CHANGES_REQUESTED** [review]

File: LICENSE
Comment (tbrent) [reply LICENSE:1]: Jeremy got it approved by Gertler a while back. I can't find it written down anywhere though.

File: contracts/ownership/Ownable.sol
Comment (tbrent) [reply contracts/ownership/Ownable.sol:67]: When we want RSV to be not controlled by anyone, i.e. it can't be paused.

File: contracts/rsv/Reserve.sol
Comment (tbrent) [reply contracts/rsv/Reserve.sol:72]: It doesn't fail there

File: contracts/rsv/Reserve.sol
Comment (tbrent) [reply contracts/rsv/Reserve.sol:194]: Wiping is gone in the new version

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:38]: Yep, staying true.

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:79]: You and Matt should chat on names, he likes them shorter than you do. I'm agnostic

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:118]: We should only use one, I agree. Unfortunately, Solidity documentation says use trailing underscores while in practice everyone does leading.

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:234]: different in new design

File: contracts/Manager.sol
Comment (tbrent) [reply contracts/Manager.sol:299]: Other functions will already revert if there is no basket. Unpause is the only one that wouldn't that otherwise should.

Also we no longer use the zero address in this way.

File: contracts/Proposal.sol
Comment (tbrent) [reply contracts/Proposal.sol:19]: Yep, nice idea. That's the design Matt landed on too.

**Review by tbrent — COMMENTED** [review]
I've read over each of the comments. For almost all of them I either replied with an explanation, or indicated with a thumbs up that it's a good point. Usually in these cases it is already changed in the next version of the code.

File: LICENSE
Comment (fiddlemath) [reply LICENSE:1]: I have this approval from Gertler in my Slack.

File: contracts/Basket.sol
Comment (fiddlemath) [reply contracts/Basket.sol:18]: Strongly disagree! I think that's up there with Hungarian notation, as a way of making code harder to read.

File: contracts/Basket.sol
Comment (fiddlemath) [reply contracts/Basket.sol:52]: much better names in current version.

File: contracts/Basket.sol
Comment (fiddlemath) [reply contracts/Basket.sol:6]: Agreed; this is improved in current version.

File: contracts/Proposal.sol
Comment (fiddlemath) [reply contracts/Proposal.sol:19]: Yep. This is true in `thursday`.

File: contracts/Proposal.sol
Comment (fiddlemath) [reply contracts/Proposal.sol:37]: Done.

File: contracts/Proposal.sol
Comment (fiddlemath) [reply contracts/Proposal.sol:56]: Done.

File: contracts/Proposal.sol
Comment (fiddlemath) [reply contracts/Proposal.sol:93]: Done.

File: contracts/Proposal.sol
Comment (fiddlemath) [reply contracts/Proposal.sol:94]: Caught in refactoring. See `thursday`, or later branches.

File: contracts/Manager.sol
Comment (fiddlemath) [reply contracts/Manager.sol:157]: It's inlined now.

File: contracts/Manager.sol
Comment (fiddlemath) [reply contracts/Manager.sol:407]: Done!

**Comment by fiddlemath** [comment]
I've gone through the reviews myself, and resolved all conversations that are no longer applicable. These leaves only 3 or 4.

**Pull request merged**
